src/test/java/com/squareup/javapoet/CodeBlockTest.java

```diff
@@ -81,27 +83,82 @@
       assertThat(exp).hasMessage("$$, $>, $<, $[ and $] may not have an index");
     }
   }
-  
+
   @Test public void nameFormatCanBeIndexed() {
     CodeBlock block = CodeBlock.builder().add("$1N", "taco").build();
     assertThat(block.toString()).isEqualTo("taco");
   }
-  
+
   @Test public void literalFormatCanBeIndexed() {
     CodeBlock block = CodeBlock.builder().add("$1L", "taco").build();
     assertThat(block.toString()).isEqualTo("taco");
   }
-  
+
   @Test public void stringFormatCanBeIndexed() {
     CodeBlock block = CodeBlock.builder().add("$1S", "taco").build();
     assertThat(block.toString()).isEqualTo("\"taco\"");
   }
-  
+
   @Test public void typeFormatCanBeIndexed() {
     CodeBlock block = CodeBlock.builder().add("$1T", String.class).build();
     assertThat(block.toString()).isEqualTo("java.lang.String");
   }
-  
+
+  @Test public void simpleNamedArgument() {
+    Map<String, Object> map = new HashMap<>();
+    map.put("text", "taco");
+    CodeBlock block = CodeBlock.builder().addNamed("$text:S", map).build();
+    assertThat(block.toString()).isEqualTo("\"taco\"");
+  }
+
+  @Test public void repeatedNamedArgument() {
+    Map<String, Object> map = new HashMap<>();
+    map.put("text", "tacos");
+    CodeBlock block = CodeBlock.builder()
+        .addNamed("\"I like \" + $text:S + \". Do you like \" + $text:S + \"?\"", map)
+        .build();
+    assertThat(block.toString()).isEqualTo(
+        "\"I like \" + \"tacos\" + \". Do you like \" + \"tacos\" + \"?\"");
+  }
+
+  @Test public void namedAndNoArgFormat() {
+    Map<String, Object> map = new HashMap<>();
+    map.put("text", "tacos");
+    CodeBlock block = CodeBlock.builder()
+        .addNamed("$>\n$text:L for $$3.50", map).build();
+    assertThat(block.toString()).isEqualTo("\n  tacos for $3.50");
+  }
+
+  @Test public void missingNamedArgument() {
+    try {
+      Map<String, Object> map = new HashMap<>();
```

Comment:
supernit: HashMap is [dead](https://publicobject.com/2016/02/08/linkedhashmap-is-always-better-than-hashmap/) to me, and it should be dead to you too. LinkedHashMap is the only way to live.